File: earthsync/android/app/src/main/kotlin/com/example/earthsync/BinauralPlugin.kt
package com.example.earthsync

import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioTrack
import android.media.audiofx.Equalizer
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlin.math.sin

class BinauralPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private var leftTrack: AudioTrack? = null
    private var rightTrack: AudioTrack? = null
    private var equalizer: Equalizer? = null
    private val sampleRate = 44100
    private val bufferSize = AudioTrack.getMinBufferSize(sampleRate, AudioFormat.CHANNEL_OUT_STEREO, AudioFormat.ENCODING_PCM_16BIT)
    private var volume = 1.0f

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "earthsync/binaural")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setFrequency" -> {
                val frequency = call.argument<Double>("frequency") ?: 7.83
                stopTracks()
                startTracks(frequency)
                result.success(null)
            }
            "setVolume" -> {
                volume = (call.argument<Double>("volume") ?: 1.0).toFloat().coerceIn(0.0f, 1.0f)
                leftTrack?.setVolume(volume)
                rightTrack?.setVolume(volume)
                result.success(null)
            }
            "setBackground" -> {
                // Background handling managed by just_audio_background
                result.success(null)
            }
            "setEqualizer" -> {
                val settings = call.arguments as Map<String, Int>
                val bass = settings["bass"] ?: 50
                val mid = settings["mid"] ?: 50
                val treble = settings["treble"] ?: 50
                configureEqualizer(bass, mid, treble)
                result.success(null)
            }
            "dispose" -> {
                stopTracks()
                equalizer?.release()
                equalizer = null
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun startTracks(frequency: Double) {
        val baseFreq = 200.0
        leftTrack = createTrack(baseFreq)
        rightTrack = createTrack(baseFreq + frequency)
        leftTrack?.setVolume(volume)
        rightTrack?.setVolume(volume)
        equalizer = Equalizer(0, leftTrack!!.audioSessionId).apply { enabled = true }
        leftTrack?.play()
        rightTrack?.play()
    }

    private fun createTrack(freq: Double): AudioTrack {
        val track = AudioTrack(
            AudioManager.STREAM_MUSIC, sampleRate, AudioFormat.CHANNEL_OUT_MONO,
            AudioFormat.ENCODING_PCM_16BIT, bufferSize, AudioTrack.MODE_STREAM
        )
        val buffer = ShortArray(bufferSize)
        Thread {
            var t = 0.0
            while (track.state == AudioTrack.STATE_INITIALIZED) {
                for (i in buffer.indices) {
                    buffer[i] = (sin(2 * Math.PI * freq * t / sampleRate) * Short.MAX_VALUE).toInt().toShort()
                    t += 1.0
                }
                track.write(buffer, 0, buffer.size)
            }
        }.start()
        return track
    }

    private fun configureEqualizer(bass: Int, mid: Int, treble: Int) {
        equalizer?.let {
            val bands = it.numberOfBands
            for (i in 0 until bands) {
                val level = when (i) {
                    0 -> bass - 50 // Bass band
                    1 -> mid - 50  // Mid band
                    else -> treble - 50 // Treble band
                }
                it.setBandLevel(i.toShort(), (level * 100).toShort()) // Scale to -5000 to 5000
            }
        }
    }

    private fun stopTracks() {
        leftTrack?.stop()
        leftTrack?.release()
        rightTrack?.stop()
        rightTrack?.release()
        leftTrack = null
        rightTrack = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        stopTracks()
    }
}

File: earthsync/ios/Runner/BinauralPlugin.swift
import Flutter
import UIKit
import AVFoundation

@objc class BinauralPlugin: NSObject, FlutterPlugin {
    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "earthsync/binaural", binaryMessenger: registrar.messenger())
        let instance = BinauralPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    private var engine: AVAudioEngine?
    private var leftNode: AVAudioPlayerNode?
    private var rightNode: AVAudioPlayerNode?
    private var equalizer: AVAudioUnitEQ?
    private var volume: Float = 1.0

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setFrequency":
            let frequency = call.arguments as? [String: Double] ?? ["frequency": 7.83]
            stopAudio()
            startAudio(frequency: frequency["frequency"]!)
            result(nil)
        case "setVolume":
            volume = Float(call.arguments as? [String: Double] ?? ["volume": 1.0])["volume"]!.clamped(to: 0.0...1.0)
            leftNode?.volume = volume
            rightNode?.volume = volume
            result(nil)
        case "setBackground":
            // Background handled by just_audio_background
            result(nil)
        case "setEqualizer":
            let settings = call.arguments as? [String: Int] ?? ["bass": 50, "mid": 50, "treble": 50]
            configureEqualizer(bass: settings["bass"]!, mid: settings["mid"]!, treble: settings["treble"]!)
            result(nil)
        case "dispose":
            stopAudio()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func startAudio(frequency: Double) {
        engine = AVAudioEngine()
        leftNode = AVAudioPlayerNode()
        rightNode = AVAudioPlayerNode()
        equalizer = AVAudioUnitEQ(numberOfBands: 3)

        let baseFreq = 200.0
        let sampleRate = 44100.0
        let bufferSize = 4096
        let leftBuffer = AVAudioPCMBuffer(pcmFormat: AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 1)!, frameCapacity: UInt32(bufferSize))!
        let rightBuffer = AVAudioPCMBuffer(pcmFormat: AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 1)!, frameCapacity: UInt32(bufferSize))!

        for i in 0..<bufferSize {
            let t = Double(i) / sampleRate
            leftBuffer.floatChannelData![0][i] = Float(sin(2 * Double.pi * baseFreq * t))
            rightBuffer.floatChannelData![0][i] = Float(sin(2 * Double.pi * (baseFreq + frequency) * t))
        }
        leftBuffer.frameLength = UInt32(bufferSize)
        rightBuffer.frameLength = UInt32(bufferSize)

        engine!.attach(leftNode!)
        engine!.attach(rightNode!)
        engine!.attach(equalizer!)
        engine!.connect(leftNode!, to: equalizer!, format: leftBuffer.format)
        engine!.connect(rightNode!, to: equalizer!, format: rightBuffer.format)
        engine!.connect(equalizer!, to: engine!.mainMixerNode, format: leftBuffer.format)

        try! engine!.start()
        leftNode!.volume = volume
        rightNode!.volume = volume
        leftNode!.play()
        rightNode!.play()

        leftNode!.scheduleBuffer(leftBuffer, at: nil, options: .loops)
        rightNode!.scheduleBuffer(rightBuffer, at: nil, options: .loops)
    }

    private func configureEqualizer(bass: Int, mid: Int, treble: Int) {
        equalizer?.bands[0].gain = Float(bass - 50) // Bass
        equalizer?.bands[1].gain = Float(mid - 50)   // Mid
        equalizer?.bands[2].gain = Float(treble - 50) // Treble
    }

    private func stopAudio() {
        leftNode?.stop()
        rightNode?.stop()
        engine?.stop()
        engine = nil
        leftNode = nil
        rightNode = nil
        equalizer = nil
    }
}

extension Comparable {
    func clamped(to limits: ClosedRange<Self>) -> Self {
        return min(max(self, limits.lowerBound), limits.upperBound)
    }
}